File: sdk/python/smcp/envelope.py
import json
from datetime import datetime, timezone
from typing import Dict, Any

from .crypto import Ed25519Key

def parse_iso8601_to_unix(timestamp_iso: str) -> int:
    """
    Parse an ISO 8601 UTC timestamp to Unix seconds.
    Expects format: YYYY-MM-DDTHH:MM:SS.sssZ or similar valid ISO string.
    """
    # Replace Z with +00:00 for strict fromisoformat parsing in standard datetime
    clean_iso = timestamp_iso.replace('Z', '+00:00')
    dt = datetime.fromisoformat(clean_iso)
    return int(dt.timestamp())

def create_canonical_message(security_token: str, payload: Dict[str, Any], timestamp_unix: int) -> bytes:
    """
    Construct democratic, deterministic message byte sequence for signing/verification.
    
    Args:
        security_token: JWT string representation.
        payload: Standard MCP JSON-RPC object dictionary.
        timestamp_unix: Unix integer tracking creation seconds.
        
    Returns:
        bytes: UTF-8 encoded JSON matching RFC requirements.
    """
    message = {
        "security_token": security_token,
        "payload": payload,
        "timestamp": timestamp_unix
    }
    
    # Serialize to JSON with sorted keys, no whitespace, and ensure ASCII is false
    # to support accurate UTF-8 canonical transmission of payload data natively.
    canonical_json = json.dumps(
        message,
        sort_keys=True,
        separators=(',', ':'),
        ensure_ascii=False
    )
    
    return canonical_json.encode('utf-8')

def create_smcp_envelope(security_token: str, mcp_payload: Dict[str, Any], private_key: Ed25519Key) -> Dict[str, Any]:
    """
    Wrap an MCP JSON-RPC payload in an SMCP Security Envelope v1.
    
    Args:
        security_token: The client's JWT allocated from attestation.
        mcp_payload: The raw standard `tools/call` JSON dictionary.
        private_key: The client's ephemeral Ed25519 keypair for signing.
        
    Returns:
        Dict: A comprehensive SMCP Envelope satisfying all RFC constraints.
    """
    utc_now = datetime.now(timezone.utc)
    timestamp_iso = utc_now.isoformat(timespec='milliseconds').replace('+00:00', 'Z')
    timestamp_unix = int(utc_now.timestamp())
    
    canonical_bytes = create_canonical_message(
        security_token=security_token,
        payload=mcp_payload,
        timestamp_unix=timestamp_unix
    )
    
    signature_b64 = private_key.sign_base64(canonical_bytes)
    
    return {
        "protocol": "smcp/v1",
        "security_token": security_token,
        "signature": signature_b64,
        "payload": mcp_payload,
        "timestamp": timestamp_iso
    }

File: sdk/typescript/src/envelope.ts
import { Ed25519Key } from './crypto';

/**
 * Standard standard json-rpc definition constraints matching MCP payloads.
 */
export interface McpPayload {
    jsonrpc: string;
    id: string | number;
    method: string;
    params?: Record<string, any>;
}

/**
 * High-fidelity wrapping envelope structure capturing security rules.
 */
export interface SmcpEnvelope {
    protocol: string;
    security_token: string;
    signature: string;
    payload: McpPayload;
    timestamp: string;
}

/**
 * Construct democratic, deterministic message byte sequence for signing/verification.
 * Following RFC canonical definitions.
 */
export function createCanonicalMessage(
    securityToken: string,
    payload: McpPayload,
    timestampUnix: number
): Uint8Array {
    const message = {
        security_token: securityToken,
        payload: payload,
        timestamp: timestampUnix,
    };

    // Stringify with sorted keys and no whitespaces
    const canonicalJson = stableStringify(message);

    // Encode into utf-8 arrays for signing boundaries
    return new TextEncoder().encode(canonicalJson);
}

/**
 * Helper to stringify an object with keys sorted alphabetically.
 */
function stableStringify(obj: any): string {
    if (obj === null || typeof obj !== 'object') {
        return JSON.stringify(obj);
    }

    if (Array.isArray(obj)) {
        const len = obj.length;
        let res = '[';
        for (let i = 0; i < len; i++) {
            res += (i ? ',' : '') + stableStringify(obj[i]);
        }
        res += ']';
        return res;
    }

    const keys = Object.keys(obj).sort();
    let res = '{';
    for (let i = 0; i < keys.length; i++) {
        const key = keys[i];
        if (obj[key] !== undefined) {
            res += (i ? ',' : '') + JSON.stringify(key) + ':' + stableStringify(obj[key]);
        }
    }
    res += '}';
    return res;
}

/**
 * Wrap an MCP JSON-RPC payload in an SMCP Security Envelope v1.
 */
export async function createSmcpEnvelope(
    securityToken: string,
    mcpPayload: McpPayload,
    privateKey: Ed25519Key
): Promise<SmcpEnvelope> {
    const now = new Date();

    // Create precisely formatted timestamps
    const timestampIso = now.toISOString();
    // Math.floor required to resolve unix equivalent properly
    const timestampUnix = Math.floor(now.getTime() / 1000);

    const canonicalBytes = createCanonicalMessage(
        securityToken,
        mcpPayload,
        timestampUnix
    );

    const signatureB64 = await privateKey.signBase64(canonicalBytes);

    return {
        protocol: 'smcp/v1',
        security_token: securityToken,
        signature: signatureB64,
        payload: mcpPayload,
        timestamp: timestampIso,
    };
}
